extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerMemoryParameters.java

```diff
@@ -309,8 +316,12 @@ public static WorkerMemoryParameters createInstance(
         )
     );
 
-    // Apportion max frames to all processors equally, then subtract one to account for an output frame.
-    final int superSorterMaxChannelsPerProcessor = maxNumFramesForSuperSorter / superSorterMaxActiveProcessors - 1;
+    final int isSmallWorker = usableMemoryInJvm < SMALL_WORKER_CAPACITY_THRESHOLD_BYTES ? 1 : 0;
+    // Apportion max frames to all processors equally, then subtract one to account for an output frame and one to account
+    // for the durable storage's output frame in the supersorter
```

Comment:
* Why does durable storage have an extra output frame?
* Also why is the subtraction valid for small heaps and not bigger heaps. 
* Please add a test case for `WorkerMemoryParametersTest` throwing NotEnoughtMemoryFault for a smaller heap.

Reply:
Updated with tests and the explanation in the comments. Regarding 2nd point, I have mentioned that we need to be conservative with the heap space in case of smaller heaps hence the subtraction is going on there and not in the case of larger heaps where we can have more parallelization without the risk of OOMing due to these temporary allocator factories.